platforms/core-execution/daemon-server-worker/src/main/java/org/gradle/workers/internal/WorkerProcessIsolationProblemsServiceProvider.java

```diff
@@ -0,0 +1,188 @@
+import org.gradle.tooling.internal.provider.serialization.DefaultPayloadClassLoaderRegistry;
+import org.gradle.tooling.internal.provider.serialization.PayloadSerializer;
+import org.gradle.tooling.internal.provider.serialization.WellKnownClassLoaderRegistry;
+
+import javax.annotation.Nonnull;
+
+public class WorkerProcessIsolationProblemsServiceProvider implements ServiceRegistrationProvider {
+
+    void configure(ServiceRegistration serviceRegistration) {
+        serviceRegistration.add(FileLookup.class, DefaultFileLookup.class);
+    }
+
+    @Nonnull
+    @Provides
+    PayloadSerializer createPayloadSerializer() {
+        ClassLoaderCache classLoaderCache = new ClassLoaderCache();
+
+        ClassLoader parent = WorkerAction.class.getClassLoader();
+        FilteringClassLoader.Spec filterSpec = new FilteringClassLoader.Spec();
+        FilteringClassLoader modelClassLoader = new FilteringClassLoader(parent, filterSpec);
+
+        return new PayloadSerializer(
+            new WellKnownClassLoaderRegistry(
+                new DefaultPayloadClassLoaderRegistry(
+                    classLoaderCache,
+                    new ModelClassLoaderFactory(modelClassLoader))));
+    }
+
+    @Provides
+    PatternSpecFactory createPatternSpecFactory(ListenerManager listenerManager) {
+        PatternSpecFactory patternSpecFactory = PatternSpecFactory.INSTANCE;
+        listenerManager.addListener(patternSpecFactory);
+        return patternSpecFactory;
+    }
+
+    @Provides
+    DirectoryFileTreeFactory createDirectoryFileTreeFactory(Factory<PatternSet> patternSetFactory, FileSystem fileSystem) {
+        return new DefaultDirectoryFileTreeFactory(patternSetFactory, fileSystem);
+    }
+
+    @Provides
+    Factory<PatternSet> createPatternSetFactory(final PatternSpecFactory patternSpecFactory) {
+        return PatternSets.getPatternSetFactory(patternSpecFactory);
+    }
+
+    @Provides
+    ClassLoaderHierarchyHasher createClassLoaderHierarchyHasher() {
+        return classLoader -> {
+            throw new UnsupportedOperationException();
+        };
+    }
+
+    @Provides
+    IsolatableSerializerRegistry createIsolatableSerializerRegistry(
+        ManagedFactoryRegistry managedFactoryRegistry,
+        ClassLoaderHierarchyHasher classLoaderHierarchyHasher
+    ) {
+        return new IsolatableSerializerRegistry(classLoaderHierarchyHasher, managedFactoryRegistry);
+    }
+
+    @Provides
+    IsolatableFactory createIsolatableFactory(
+        ManagedFactoryRegistry managedFactoryRegistry,
+        ClassLoaderHierarchyHasher classLoaderHierarchyHasher
+    ) {
+        return new DefaultIsolatableFactory(classLoaderHierarchyHasher, managedFactoryRegistry);
+    }
+
+    @Provides
+    PropertyHost createPropertyHost() {
+        return PropertyHost.NO_OP;
+    }
+
+    @Provides
+    CrossBuildInMemoryCacheFactory createCrossBuildInMemoryCacheFactory(ListenerManager listenerManager) {
```

Comment:
This does not seem right to me. The concept of "cross build" does not make sense in a worker context. Workers are not aware of build lifecycle

All you should need is a `ClassCacheFactory`, which should already be available in worker processes.

Reply:
Seems to be not needed at all. thx